CMakeLists.txt

```diff
@@ -66,11 +66,10 @@ zephyr_compile_definitions(
   __ZEPHYR__=1
 )
 
-if(NOT CONFIG_NO_OPTIMIZATIONS)
-zephyr_compile_definitions(
-  _FORTIFY_SOURCE=2
-)
-endif()
+# @Intent: Set compiler flags related to security, which is specific to each compiler
+# @config in CONFIG_STACK_CANARIES   optional : Stack canaries can detect stack overflows
+# @config in CONFIG_NO_OPTIMIZATIONS optional : Optimizations may affect security
```

Comment:
About the granularity or abstraction level of the intents.

It must be possible for a user to read the intent documentation, and be able to roughly predict how it affects the build without reading the toolchain-specific build scripts.

If the intents are too high-level, like "use security", then we
won't have this ability.

For this reason, and for other reasons that I won't go into
unless necessary, we need to split the intent "use security"
into, for instance; "use stack canaries", and "detect buffer overflow
in libc functions".

Reply:
Alright, good feedback. I agree about the documentation.

So you would like something like `toolchain_cc_security_stack_canaries` or simply `toolchain_cc_stack_canaries`?

I'm fine with splitting security into further pieces. 
However, the choice of level of abstraction has pros and cons:

- Pro for high-level: It permits commerical toolchains to enable checks that are not nessecarily part of upstream Zephyr. E.g. Zephyr does not have retpoline. Oticon may want it. But without a generic `toolchain_cc_security` where would we place it? There would be only two options (to misuse) `toolchain_cc_security_stack_canaries` or `toolchain_cc_security_fortify`. Other cases will appear beyond the security category. 
- Pro for high-level: Reduced-patch rate and lower noise of `/CMakeLists.txt`.
- Con for high-level: Loss of detail at `/CMakeLists.txt` (inevitable IMHO now that we are looking to broadening the toolchain horizon).
- Meh for low-level: Errors will be emitted if an existing toolchain does not support a newly added intent-macro. Errors are good I guess.

As you found, Fortify is a GNU thing. So it should not be a requirement for all toolchains to implement `toolchain_cc_security_fortify()`.

Reply:
> So you would like something like toolchain_cc_security_stack_canaries or simply toolchain_cc_stack_canaries?

Yes.

> I'm fine with splitting security into further pieces.

Great.

> Oticon may want it. But without a generic toolchain_cc_security where would we place it? 

retpoline's are not a good example as they will likely be supported in the near future, and should therefore be placed in toolchain_cc_security_retpoline(). But I get your point, where do we put changes that a toolchain needs / wants to make that only make sense to that particular toolchain?

It's not obvious to me where these should be put, I think it depends on how that toolchain feature affects the build. A good example might help clarify. But perhaps they would be appropriate in cmake/toolchain/$TOOLCHAIN/*.cmake.